arrows/vxl/pixel_feature_extractor.cxx

```diff
@@ -173,49 +246,104 @@ ::filter( kwiver::vital::image_container_sptr input_image )
     if( enable_gray )
     {
       // TODO consider vil_convert_to_grey_using_rgb_weighting
-      filtered_images.push_back( vil_convert_to_grey_using_average( vxl_image ) );
+      auto vxl_gray_sptr =
+        vil_convert_to_grey_using_average(
+          vxl::image_container::vital_to_vxl( input_image->get_image() ) );
+      auto vxl_gray = vil_convert_cast( pix_t(), vxl_gray_sptr );
+      filtered_images.push_back( vxl_gray );
     }
   }
 
   if( enable_color_commonality )
   {
     // 1 channel
-    auto color_commonality = color_commonality_filter->filter( input_image );
-    filtered_images.push_back(
-        vxl::image_container::vital_to_vxl( color_commonality->get_image() ) );
+    auto color_commonality = convert_to_typed_vil_image_view< pix_t >(
+      color_commonality_filter->filter( input_image ) );
+
+    filtered_images.push_back( color_commonality );
   }
   if( enable_high_pass_box )
   {
-    auto high_pass_box = high_pass_box_filter->filter( input_image );
-    // 2 channels
-    filtered_images.push_back(
-        vxl::image_container::vital_to_vxl( high_pass_box->get_image() ) );
+    auto high_pass_box = convert_to_typed_vil_image_view< pix_t >(
+      high_pass_box_filter->filter( input_image ) );
+
+    // Legacy BurnOut models expect these channels to be incorrectly ordered
+    // TODO Remove this code when we no longer need to train models using
+    // legacy code
+    auto first_plane = vil_plane( high_pass_box, 0 );
+    auto second_plane = vil_plane( high_pass_box, 1 );
+    auto temp = vil_copy_deep( first_plane );
+    first_plane.deep_copy( second_plane );
+    second_plane.deep_copy( temp );
+
+    // 3 channels
+    filtered_images.push_back( high_pass_box );
   }
   if( enable_high_pass_bidir )
   {
-    auto high_pass_bidir = high_pass_bidir_filter->filter( input_image );
-    // 2 channels
-    filtered_images.push_back(
-        vxl::image_container::vital_to_vxl( high_pass_bidir->get_image() ) );
+    auto high_pass_bidir = convert_to_typed_vil_image_view< pix_t >(
+      high_pass_bidir_filter->filter( input_image ) );
+    // 3 channels
+    filtered_images.push_back( high_pass_bidir );
+  }
+
+  kwiver::vital::image_container_sptr variance_container;
+
+  if( enable_average || enable_normalized_variance )
+  {
+    // This is only used internally and isn't externally configurable
+    vital::config_block_sptr convert_config =
+      vital::config_block::empty_config();
+    convert_config->set_value( "single_channel", true );
+    convert_filter->set_configuration( convert_config );
+
+    auto grayscale = convert_filter->filter( input_image );
```

Comment:
Constructing an algorithm *every time* seems gratuitous. Do we even *need* the algorithm here, or could we just inline this functionality? (If we do need the algorithm, would it make sense to make it a member rather than a local?)

Reply:
Yes the algorithm should be configured elsewhere. However, I think having an algorithm rather than converting to and from an kwiver image container explicitly makes cleaner.

Reply:
Actually, now that you mention it... this seems to be the only reason `variance_container` *is* a container and not a VIL image? But `convert_filter` is definitely a VXL algorithm, which means we know the variance image *is* a VIL image. Wouldn't it actually simplify things to inline the grayscale conversion and just have a VIL `double` image instead of `variance_container`?

Reply:
For that matter, even leaving this as a filter, it seems like it would be simpler to immediately convert the output to a VIL image... do it once, here, rather than in two other places (and potentially twice!).

Reply:
I updated this to inline and it failed the unit tests turned out to do an unnecessary cast to `byte`. So I updated the expected data.